### PR TITLE
perf(robot-server): move large response serialization to worker thread

### LINKED ISSUE #9133: bug: Protocol execution stalls when unrelated tasks hog the main thread's async event loop
# Overview

This ticket tracks a specific theory for why an OT-2 running the v5.0.0 beta would be unusually janky (jittery, stuttery, sluggish, slow) when executing a protocol.

# The user-reported problem

> It was extremely slow. I ran a protocol involving a bunch of mixing and transferring and the robot was frequently pausing and much less [smooth] than usual.

(See Matt for a copy of this user's protocol.)

My own testing shows that it gets much worse when you spam the robot with HTTP requests, especially `GET /protocols` or `GET /runs`.

# The technical theory

APIv2 protocols are written in plain old synchronous Python, without any `async` or `await`. We execute them in a separate thread, so they *should* be mostly isolated, performance-wise, from the HTTP server doing HTTP things.

However, even though the protocol isn't *directly* participating in the main thread's event loop, it will *effectively* still need to wait for that event loop to be free at certain times:

* When moving the hardware. At a low level, hardware control is async in the main thread's event loop. If the low-level async hardware control gets stuck waiting behind something else that's hogging the event loop, the protocol thread will feel it, because it can't proceed until the hardware command has completed.
* When reporting the protocol's activity to the `ProtocolEngine`. The `ProtocolEngine` is owned by the main thread, shared with the HTTP API, and is generally async. Currently, when a command is executed in the protocol thread, the APIv2 internals schedule an async task to run in the main thread and update the `ProtocolEngine`. But, they wait until that update has completed before allowing the protocol to continue. The update itself should be quick, but it may be stuck waiting behind something else hogging the event loop.

Some things are known or suspected to hog the event loop:

* HTTP response serialization and validation, especially for unbounded responses like `GET /protocols/{id}` and `GET /runs/{id}`.
* Parsing and filesystem queries, like in #6951.
* Inefficient data structure access.
* Possibly, other things that we're missing. `robot-server` is a big process, responsible for a lot.

# Acceptance criteria

This ticket can be closed when the robot *doesn't* visibly stutter or slow down when an ill-behaved task accidentally hogs the event loop.

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This is a~n experimental~ PR that moves large response serialization to worker threads.

## Changelog

- perf(robot-server): move large response serialization to worker thread
    - Basically anything that returns one or more `Protocol`, `Run`, or `Command` resource models

## Review requests

- Does this impact run performance in 5.x?

## Risk assessment

Low

### CODE REVIEW

**Comment by mcous** [comment]
In my robot testing using a very simple "large" protocol:

```py
from opentrons import types

metadata = {
    "protocolName": "Testosaur",
    "author": "Opentrons <engineering@opentrons.com>",
    "description": 'A variant on "Dinosaur" for testing',
    "source": "Opentrons Repository",
    "apiLevel": "2.0",
}


REPEAT_COUNT = 500


def run(ctx):
    ctx.home()
    tr = ctx.load_labware("opentrons_96_tiprack_300ul", 1)
    right = ctx.load_instrument("p300_single", types.Mount.RIGHT, [tr])
    lw = ctx.load_labware("corning_96_wellplate_360ul_flat", 2)

    tip = tr.wells()[0].top()
    source_well = lw.wells()[0].top()
    dest_well = lw.wells()[1].top()

    for i in range(REPEAT_COUNT):
        right.pick_up_tip(tip)
        right.aspirate(10, source_well)
        right.dispense(10, dest_well)
        right.drop_tip(tip)
```

Before this PR:

- Full "commands" will complete without pause thanks to the ThreadManager
- A badly timed HTTP request can block robot movements between commands

After this PR:

- Most HTTP requests by the app during a run won't noticeably block robot movements between commands
    - However, this only applies to `/protocols` and `/runs` endpoints
- Anything sync in the non-serialization phase of the request handlers can (and probably does) still block robot movements between commands, it's just less noticeable now

TL;DR: clear improvement that gets us most of the way there. One last change or two needed for the RC, and probably an audit of endpoints outside `/protocols` and `/runs` to make sure they're well behaved (e.g. #6951)

**Review by b-cooper — APPROVED** [review]
Tested on robot, improves perf on large protocol 🍨
